### ISSUE TITLE
Vocabulary expansion  pre-training code for Chinese language 

### ISSUE THREAD

**Issue by abdoelsayed2016** [ymcui/Chinese-LLaMA-Alpaca#161]
Thank you for great work, 

i read all step about pretraining to Preparation: Vocabulary Expansion, i got confused becuase i want to work on arabic language which is not support by llama tokenizer .

After you collect vocab from [sentencepiece](https://github.com/google/sentencepiece) 

Did you add the vocals to the tokenizer using sentencepieces and create a new tokenizer?

And then you used the new tokenizer with the llama model for fine-tuning using [run_clm.py](https://github.com/huggingface/transformers/blob/main/examples/pytorch/language-modeling/run_clm.py) 

Can you post the hyperparameters used for rnu_clm.py and how many epochs are used?

And also, can you give more explanation about the steps that you wrote in the readme

Stage One: Fix the parameters of the transformer part of the model and only train the embedding, adapting the newly added Chinese word vectors without disturbing the original model as much as possible.
Stage Two: Use LoRA technology to add LoRA weights (adapter) to the model and train the embedding while updating LoRA parameters.

What is stage two? did you used it after pre training the model ?

**Comment by airaria** [comment]
> After you collect vocab from [sentencepiece](https://github.com/google/sentencepiece) Did you add the vocals to the tokenizer using sentencepieces and create a new tokenizer?

Yes, We create a new tokenizer by adding tokens from Chinese tokenizer to the original LLaMA tokenizer using sentencepiece.

> Can you post the hyperparameters used for rnu_clm.py and how many epochs are used?

batch size 2K, peak learning rate 2e-4, training for 1 epochs.

We use [peft](https://github.com/huggingface/peft) to add LoRA adapter to he LLaMA model, and freeze the params in encoders, only train the LoRA adapter and embed_tokens and lm_head. 

Both Stage One and Stage Two are pre-training. We train more params in Stage Two.

**Comment by abdoelsayed2016** [comment]
do you mean that in
stage one you used [run_clm.py](https://github.com/huggingface/transformers/blob/main/examples/pytorch/language-modeling/run_clm.py)
and stage two the output model from run_clm then used in  [merge_llama_with_chinese_lora.py](https://github.com/ymcui/Chinese-LLaMA-Alpaca/blob/main/scripts/merge_llama_with_chinese_lora.py)

**Comment by abdoelsayed2016** [comment]
can you give more explaination about that please

**Comment by airaria** [comment]
In stage One, the input is Original LLaMA, we use run_clm.py to pre-train, and we get Model A
In stage Two, we first load Model A, and add the LoRA adapters to it using peft library. We call it Model B.
Then we use run_clm.py to pre-train Model B. The output is Chinese-LLaMA.

**Comment by abdoelsayed2016** [comment]
So did you train run_clm.py two times?

**Comment by airaria** [comment]
> So did you train run_clm.py two times?

Yes. We run run_clm.py two times for 7B model

**Comment by abdoelsayed2016** [comment]
Could you please post the code for the stage two?

**Comment by abdoelsayed2016** [comment]
To enhance the comprehensiveness of the README, could you please adding more details about the process. This could include providing step-by-step instructions, explanations of each step, and any prerequisites or dependencies required for the process to work. Additionally, consider including examples or screenshots to illustrate the process and make it more user-friendly. By doing so, users will have a clearer understanding of how to navigate through the process and achieve the desired outcome. 

We would like to express our gratitude for your invaluable assistance, and we are interested in extending your work .

Thank you.

**Comment by ehion** [comment]
@abdoelsayed2016 totally agree with u，this project does not mention the important thing，they just give simple code to show Chinese llama。
As your problem in my view:
first you should use sentencepiece BPE algorithm to train your language and you will get a *.model file(Llama also use sentencepiece BPE and you can see the source code in transformers)，BPE just learn the subword representation and the *.model file is just index between subword and index。You can write a simple code to load llama tokenizer.model or your DIY sentencepiece model to see the index
second you need to write a simple code to merge your language words vocab(sentencepiece BPE result)and the origin llama vocab(tokenizer.model) to get the final vocab(origin llama index from 0,1,2,3,...,n and your language index, n+1,n+2,n+....)
third you can do your pretraining via run_clm.py（embedding pretraining + lora pretraining）

**Comment by abdoelsayed2016** [comment]
@ehion Ya i hope they update readme. it will help many ppls bacuase i see many issue asking same question

**Comment by airaria** [comment]
Thank you for your interest in this project. 
We will soon release the code for processing tokenizers in one day or two, and keep adding more resources. 
Please be patient.

**Comment by abdoelsayed2016** [comment]
@airaria Thank you for sharing the script for merging tokenizers. I have another question regarding the training of Llama. Have you also experienced a loss of 0 while training it on Chinese data?

and did you trainned all 7b parameter ?

**Comment by airaria** [comment]
> @airaria Thank you for sharing the script for merging tokenizers. I have another question regarding the training of Llama. Have you also experienced a loss of 0 while training it on Chinese data?
> 
> and did you trainned all 7b parameter ?

For the released model, we did not train LLaMA's full parameters; We trained LLaMA with LoRA and float16.

In our preliminary experiments, we have experimented with full-parameter training for about several hundred steps; but no loss of 0 observed.

**Comment by abdoelsayed2016** [comment]
@airaria 
in strart of the training the loss was 4.3 then it is be 0 

{'loss': 0.0, 'learning_rate': 4.8303582439699045e-05, 'epoch': 0.03}
{'loss': 0.0, 'learning_rate': 4.829654336268535e-05, 'epoch': 0.03}
{'loss': 0.0, 'learning_rate': 4.828950428567165e-05, 'epoch': 0.03}
{'loss': 0.0, 'learning_rate': 4.8282465208657955e-05, 'epoch': 0.03}

is there something wrong should i stop training ?

**Comment by airaria** [comment]
> @airaria in strart of the training the loss was 4.3 then it is be 0
> 
> {'loss': 0.0, 'learning_rate': 4.8303582439699045e-05, 'epoch': 0.03} {'loss': 0.0, 'learning_rate': 4.829654336268535e-05, 'epoch': 0.03} {'loss': 0.0, 'learning_rate': 4.828950428567165e-05, 'epoch': 0.03} {'loss': 0.0, 'learning_rate': 4.8282465208657955e-05, 'epoch': 0.03}
> 
> is there something wrong should i stop training ?

hyperparameters, please?

**Comment by iou2much** [comment]
> In stage One, the input is Original LLaMA, we use run_clm.py to pre-train, and we get Model A In stage Two, we first load Model A, and add the LoRA adapters to it using peft library. We call it Model B. Then we use run_clm.py to pre-train Model B. The output is Chinese-LLaMA.

Could you please explain for more detail?
Stage 1,  with the input is Original LLaMA , you expand the vocabulary to 49K, and use run_clm.py to pre-train. It gets Model A ?
Stage 2, add LoRA adapters to Model A, and train it with peft ? after that it gets Model B.
and then use run_clm.py to pre-train Model B ?

**Comment by steamboatid** [comment]
> Thank you for great work,
> 
> i read all step about pretraining to Preparation: Vocabulary Expansion, i got confused becuase i want to work on arabic language which is not support by llama tokenizer .
> 
> After you collect vocab from [sentencepiece](https://github.com/google/sentencepiece)
> 
> Did you add the vocals to the tokenizer using sentencepieces and create a new tokenizer?
> 
> And then you used the new tokenizer with the llama model for fine-tuning using [run_clm.py](https://github.com/huggingface/transformers/blob/main/examples/pytorch/language-modeling/run_clm.py)
> 
> Can you post the hyperparameters used for rnu_clm.py and how many epochs are used?
> 
> And also, can you give more explanation about the steps that you wrote in the readme
> 
> Stage One: Fix the parameters of the transformer part of the model and only train the embedding, adapting the newly added Chinese word vectors without disturbing the original model as much as possible. Stage Two: Use LoRA technology to add LoRA weights (adapter) to the model and train the embedding while updating LoRA parameters.
> 
> What is stage two? did you used it after pre training the model ?

I second to @abdoelsayed2016, 
please provide detailed steps for stage 1 and 2.

thanks

**Comment by Ashwani-Dangwal** [comment]
> > So did you train run_clm.py two times?
> 
> Yes. We run run_clm.py two times for 7B model

@airaria You mentioned that you used two stages for pre trsining just for the 7B model. Then how were the rest of the models trained. Please show some light on this.
